client/src/di/Preprocess.ts

```diff
@@ -0,0 +1,201 @@
+ *
+ * This function iterates over the spans of the paragraph to determine the
+ * minimum and maximum character offsets, effectively defining the paragraph's
+ * position within the document's content.
+ *
+ * Edge cases:
+ * - If the paragraph has no spans, returns `[-1, -1]`.
+ *
+ * @param paragraph - The paragraph for which to compute the offset range.
+ * @returns A tuple `[minOffset, maxOffset]` representing the start and end
+ *          character positions of the paragraph in the document.
+ */
+function getParagraphOffsetRange(paragraph: Paragraph): [number, number] {
+  let minOffset = Number.MAX_SAFE_INTEGER;
+  let maxOffset = -1;
+
+  for (const sp of paragraph.spans) {
+    if (sp.offset < minOffset) {
+      minOffset = sp.offset;
+    }
+    const end = sp.offset + sp.length - 1;
+    if (end > maxOffset) {
+      maxOffset = end;
+    }
+  }
+
+  // Handle paragraphs with no spans
+  if (minOffset === Number.MAX_SAFE_INTEGER) {
+    minOffset = -1;
+    maxOffset = -1;
+  }
+
+  return [minOffset, maxOffset];
+}
+
+/**
+ * Computes the character offset range for a given document line.
+ *
+ * This function determines the minimum and maximum character offsets
+ * by iterating over the spans of the line, defining its position within
+ * the document's content.
+ *
+ * Edge cases:
+ * - If the line has no spans, returns `[-1, -1]`.
+ *
+ * @param line - The line for which to compute the offset range.
+ * @returns A tuple `[minOffset, maxOffset]` representing the start and end
+ *          character positions of the line in the document.
+ */
+function getLineOffsetRange(line: Line): [number, number] {
+  let minOffset = Number.MAX_SAFE_INTEGER;
+  let maxOffset = -1;
+
+  for (const sp of line.spans) {
+    if (sp.offset < minOffset) {
+      minOffset = sp.offset;
+    }
+    const end = sp.offset + sp.length - 1;
+    if (end > maxOffset) {
+      maxOffset = end;
+    }
+  }
+
+  // Handle lines with no spans
+  if (minOffset === Number.MAX_SAFE_INTEGER) {
+    minOffset = -1;
+    maxOffset = -1;
+  }
+
+  return [minOffset, maxOffset];
+}
+
+/**
+ * Creates a list of `Region` objects for each page in the document,
+ * grouping content based on paragraph bounding regions rather than line adjacency.
+ *
+ * This function processes the document’s paragraphs and associates them with
+ * their respective pages. Each paragraph is analyzed to compute its bounding
+ * polygon, character offset range, and word index range. The resulting regions
+ * provide a structured representation of paragraph-based segmentation.
+ *
+ * @param di - The Document Intelligence response containing analyzed paragraphs and pages.
+ */
+export function createPerPageRegions(di: DocIntResponse) {
+  const paragraphs = di.analyzeResult.paragraphs ?? [];
+  const pages = di.analyzeResult.pages ?? [];
+
+  for (const page of pages) {
+    const pageNumber = page.pageNumber;
+    const lines = page.lines || [];
```

Comment:
I prefer `??` to `||` here (but maybe I am missing some reason why `||` might be preferable)

Reply:
You're right. Using `??` is safer and make more sense.